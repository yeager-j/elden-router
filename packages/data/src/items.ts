import { Location } from "#locations";
import { Questline } from "#quests";

export enum ProgressionItem {
  DECTUS_MEDALLION = "DECTUS_MEDALLION",
  ROLD_MEDALLION = "ROLD_MEDALLION",
  HALIGTREE_MEDALLION = "HALIGTREE_MEDALLION",
  PUREBLOOD_KNIGHTS_MEDAL = "PUREBLOOD_KNIGHTS_MEDAL",
  ACADEMY_GLINTSTONE_KEY = "ACADEMY_GLINTSTONE_KEY",
  CARIAN_INVERTED_STATUE = "CARIAN_INVERTED_STATUE",
}

export interface ProgressionItemMetadata {
  displayName: string;
  description: string;
}

export const ProgressionItemData: Record<
  ProgressionItem,
  ProgressionItemMetadata
> = {
  [ProgressionItem.DECTUS_MEDALLION]: {
    displayName: "Dectus Medallion",
    description: "Fort Haight, Fort Faroth",
  },
  [ProgressionItem.ROLD_MEDALLION]: {
    displayName: "Rold Medallion",
    description:
      "Melina gives this to you after you defeat Morgott, the Omen King",
  },
  [ProgressionItem.HALIGTREE_MEDALLION]: {
    displayName: "Haligtree Secret Medallion",
    description:
      "Village of the Albinaurics, Castle Sol after defeating Commander Niall",
  },
  [ProgressionItem.PUREBLOOD_KNIGHTS_MEDAL]: {
    displayName: "Pureblood Knight's Medal",
    description: "White Mask Varre's Questline",
  },
  [ProgressionItem.ACADEMY_GLINTSTONE_KEY]: {
    displayName: "Academy Glintstone Key",
    description: "Found near Glintstone Dragon Smarag",
  },
  [ProgressionItem.CARIAN_INVERTED_STATUE]: {
    displayName: "Carian Inverted Statue",
    description: "Given to you by Ranni during her questline",
  },
};

export enum ItemType {
  KEY_ITEM = "KEY_ITEM",
  WEAPON = "WEAPON",
  ARMOR = "ARMOR",
  TALISMAN = "TALISMAN",
  CONSUMABLE = "CONSUMABLE",
  SORCERY = "SORCERY",
  INCANTATION = "INCANTATION",
}

export const ItemTypeNames: Record<ItemType, string> = {
  [ItemType.WEAPON]: "Weapon",
  [ItemType.KEY_ITEM]: "Key Item",
  [ItemType.ARMOR]: "Armor",
  [ItemType.TALISMAN]: "Talisman",
  [ItemType.CONSUMABLE]: "Consumable",
  [ItemType.SORCERY]: "Sorcery",
  [ItemType.INCANTATION]: "Incantation",
};

export enum Item {
  // Weapons
  RIVERS_OF_BLOOD = "RIVERS_OF_BLOOD",
  BOLT_OF_GRANSAX = "BOLT_OF_GRANSAX",

  // Talismans
  ERDTREE_FAVOR_2 = "ERDTREE_FAVOR_2",

  // Sorceries
  COLLAPSING_STARS = "COLLAPSING_STARS",

  // Incantations
  TRIPLE_RINGS_OF_LIGHT = "TRIPLE_RINGS_OF_LIGHT",

  // Key Items
  SOMBERSTONE_MINERS_BELL_BEARING_5 = "SOMBERSTONE_MINERS_BELL_BEARING_5",
}

export interface ItemMetadata {
  displayName: string;
  itemType: ItemType;
  location: Location;
  quest?: Questline;
}

export const ItemData: Record<Item, ItemMetadata> = {
  [Item.RIVERS_OF_BLOOD]: {
    displayName: "Rivers of Blood",
    itemType: ItemType.WEAPON,
    location: Location.MOUNTAINTOPS_OF_THE_GIANTS,
  },
  [Item.BOLT_OF_GRANSAX]: {
    displayName: "Bolt of Gransax",
    itemType: ItemType.WEAPON,
    location: Location.LEYNDELL_ERDTREE_SANCTUARY,
  },
  [Item.COLLAPSING_STARS]: {
    displayName: "Collapsing Stars",
    itemType: ItemType.SORCERY,
    location: Location.CAELID_WAR_DEAD_CATACOMBS,
  },
  [Item.TRIPLE_RINGS_OF_LIGHT]: {
    displayName: "Triple Rings of Light",
    itemType: ItemType.INCANTATION,
    location: Location.ELPHAEL_BRACE_OF_THE_HALIGTREE,
  },
  [Item.SOMBERSTONE_MINERS_BELL_BEARING_5]: {
    displayName: "Somberstone Miner's Bell Bearing (5)",
    itemType: ItemType.KEY_ITEM,
    location: Location.FARUM_AZULA_POST_GODSKIN_DUO,
  },
  [Item.ERDTREE_FAVOR_2]: {
    displayName: "Erdtree's Favor +2",
    itemType: ItemType.TALISMAN,
    location: Location.LEYNDELL_ASHEN_CAPITAL,
  },
};
