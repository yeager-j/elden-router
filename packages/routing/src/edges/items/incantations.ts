import { Item, Location } from "@workspace/data";

import { EdgeData } from "@/types";

export const incantationLocations: EdgeData[] = [
  {
    from: Location.ELPHAEL_BRACE_OF_THE_HALIGTREE,
    to: Item.TRIPLE_RINGS_OF_LIGHT,
    metadata: {
      description:
        "In a sealed treasure room just under the Prayer Room that requires a Stonesword Key to unlock.",
      requirements: [],
    },
  },
];
