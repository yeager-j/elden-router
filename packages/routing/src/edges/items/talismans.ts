import { Item, Location } from "@workspace/data";

import { EdgeData } from "@/types";

export const talismanLocations: EdgeData[] = [
  {
    from: Location.LEYNDELL_ASHEN_CAPITAL,
    to: Item.ERDTREE_FAVOR_2,
    metadata: {
      requirements: [],
    },
  },
];
