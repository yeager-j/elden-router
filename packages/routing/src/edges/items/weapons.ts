import { Item, Location } from "@workspace/data";

import { EdgeData } from "@/types";

export const weaponLocations: EdgeData[] = [
  {
    from: Location.MOUNTAINTOPS_OF_THE_GIANTS,
    to: Item.RIVERS_OF_BLOOD,
    metadata: {
      description:
        "Defeat Bloody Finger Okina when he invades in the Church of Repose",
      requirements: [],
    },
  },
  {
    from: Location.LEYNDELL_ERDTREE_SANCTUARY,
    to: Item.BOLT_OF_GRANSAX,
    metadata: {
      description:
        "This weapon is found on the giant spear lodged into the building",
      requirements: [],
    },
  },
];
