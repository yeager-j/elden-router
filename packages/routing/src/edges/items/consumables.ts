import { Item } from "@workspace/data/items";
import { Location } from "@workspace/data/locations";

import { EdgeData } from "#types";
import { convertItemLocationsToEdges } from "#utils/edge-utils";

export const somberSmithingStone4Locations: EdgeData[] =
  convertItemLocationsToEdges({
    item: Item.SOMBER_SMITHING_STONE_4,
    locations: [
      {
        location: Location.CAELID_SELLIA_CRYSTAL_TUNNEL,
        metadata: {
          requirements: {},
        },
      },
      {
        location: Location.CAELID,
        metadata: {
          requirements: {},
        },
      },
      {
        location: Location.NOKRON_ETERNAL_CITY,
        metadata: {
          requirements: {},
        },
      },
    ],
  });

export const somberSmithingStone7Locations: EdgeData[] =
  convertItemLocationsToEdges({
    item: Item.SOMBER_SMITHING_STONE_7,
    locations: [
      {
        location: Location.MT_GELMIR_VOLCANO_MANOR_POST_EIGLAY,
        metadata: {
          requirements: {},
        },
      },
      {
        location: Location.FARUM_AZULA_POST_GODSKIN_DUO,
        metadata: {
          requirements: {},
        },
      },
      {
        location: Location.DEEPROOT_DEPTHS,
        metadata: {
          requirements: {},
        },
      },
    ],
  });

export const consumableLocations: EdgeData[] = [
  ...somberSmithingStone4Locations,
  ...somberSmithingStone7Locations,
];
