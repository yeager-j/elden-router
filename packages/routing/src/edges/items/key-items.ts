import { Item } from "@workspace/data/items";
import { Location } from "@workspace/data/locations";

import { EdgeData } from "#types";

export const keyItemLocations: EdgeData[] = [
  {
    from: Location.FARUM_AZULA_POST_GODSKIN_DUO,
    to: Item.SOMBERSTONE_MINERS_BELL_BEARING_5,
    metadata: {
      description:
        "Down the elevator from the 'Beside the Great Bridge' Site of Grace in the small church",
      requirements: [],
    },
  },
];
