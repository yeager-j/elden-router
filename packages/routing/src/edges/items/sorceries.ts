import { Item } from "@workspace/data/items";
import { Location } from "@workspace/data/locations";

import { EdgeData } from "#types";

export const sorceryLocations: EdgeData[] = [
  {
    from: Location.CAELID_WAR_DEAD_CATACOMBS,
    to: Item.COLLAPSING_STARS,
    metadata: {
      description:
        "In the center of the floor of the West-most wall in the Scarlet Rot area",
      requirements: [],
    },
  },
];
