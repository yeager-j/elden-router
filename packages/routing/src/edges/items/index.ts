import { consumableLocations } from "#edges/items/consumables";
import { incantationLocations } from "#edges/items/incantations";
import { keyItemLocations } from "#edges/items/key-items";
import { sorceryLocations } from "#edges/items/sorceries";
import { talismanLocations } from "#edges/items/talismans";
import { weaponLocations } from "#edges/items/weapons";

export const allItems = [
  ...weaponLocations,
  ...talismanLocations,
  ...sorceryLocations,
  ...keyItemLocations,
  ...incantationLocations,
  ...consumableLocations,
];
