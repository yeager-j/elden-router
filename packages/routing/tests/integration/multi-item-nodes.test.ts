import { describe, expect, test } from "vitest";

import { Item, ProgressionItem } from "@workspace/data/items";

import { getPathToDestination } from "#graph";
import { BossPreference, PathSettings } from "#types";

describe("multi-location items", () => {
  const config: PathSettings = {
    bossPreference: BossPreference.MINIMAL,
    allowGlitches: true,
    acquiredItems: new Set([
      ProgressionItem.DECTUS_MEDALLION,
      ProgressionItem.ROLD_MEDALLION,
      ProgressionItem.HALIGTREE_MEDALLION,
    ]),
    completedQuestlineStages: new Set(),
    flagsEnabled: new Set(),
  };

  test("Finds the fastest way to a Somber Smithing Stone 4", () => {
    const route = getPathToDestination(Item.SOMBER_SMITHING_STONE_4, config);

    expect(route).toMatchSnapshot();
  });

  test("Finds the fastest way to a Somber Smithing Stone 7", () => {
    const route = getPathToDestination(Item.SOMBER_SMITHING_STONE_7, config);

    expect(route).toMatchSnapshot();
  });
});
