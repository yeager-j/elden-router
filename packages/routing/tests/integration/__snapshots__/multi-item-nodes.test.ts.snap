// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`multi-location items > Finds the fastest way to a Somber Smithing Stone 4 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "requirements": {},
      },
      "to": "CAELID",
    },
    {
      "from": "CAELID",
      "metadata": {
        "requirements": {},
      },
      "to": "SOMBER_SMITHING_STONE_4 - CAELID",
    },
  ],
  "reasons": [],
}
`;

exports[`multi-location items > Finds the fastest way to a Somber Smithing Stone 7 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "description": "Take the shortcut near Stormveil Castle",
        "requirements": {},
      },
      "to": "LIURNIA",
    },
    {
      "from": "LIURNIA",
      "metadata": {
        "description": "Access via Grand Lift of Dectus",
        "requirements": {
          "requiredItems": [
            "DECTUS_MEDALLION",
          ],
        },
      },
      "to": "ALTUS_PLATEAU",
    },
    {
      "from": "ALTUS_PLATEAU",
      "metadata": {
        "requirements": {},
      },
      "to": "MT_GELMIR",
    },
    {
      "from": "MT_GELMIR",
      "metadata": {
        "requirements": {},
      },
      "to": "MT_GELMIR_VOLCANO_MANOR",
    },
    {
      "from": "MT_GELMIR_VOLCANO_MANOR",
      "metadata": {
        "description": "Area after Temple of Eiglay",
        "requirements": {
          "requiredBosses": [
            "GODSKIN_NOBLE",
          ],
        },
      },
      "to": "MT_GELMIR_VOLCANO_MANOR_POST_EIGLAY",
    },
    {
      "from": "MT_GELMIR_VOLCANO_MANOR_POST_EIGLAY",
      "metadata": {
        "requirements": {},
      },
      "to": "SOMBER_SMITHING_STONE_7 - MT_GELMIR_VOLCANO_MANOR_POST_EIGLAY",
    },
  ],
  "reasons": [],
}
`;
